mlir/lib/Dialect/LLVMIR/IR/LLVMLinkerInterface.cpp

```diff
@@ -223,196 +217,147 @@ class LLVMSymbolLinkerInterface : public SymbolLinkerInterface {
-      return pair.dst;
+      StringAttr name = symbolAttr(materialized);
+      if (st.lookup(name)) {
+        StringAttr newName = getUniqueNameIn(st, name);
+        st.setSymbolName(materialized, newName);
+        toRenameUsers.push_back({op, newName});
+      }
 
-    assert(!isDeclaration(pair.src) && "expected source with body");
-    if (!pair.dst)
-      return unique(pair.src->clone(), dst);
+      st.insert(materialized);
+    }
 
-    // TODO might need more things to clone here
-    for (auto [srcRegion, dstRegion] :
-         llvm::zip(pair.src->getRegions(), pair.dst->getRegions())) {
-      dstRegion.takeBody(srcRegion);
+    for (auto &[op, newName] : toRenameUsers) {
+      if (failed(renameRemappedUsersOf(op, newName, state)))
+        return failure();
     }
 
-    return pair.dst;
+    return success();
+  }
+
+  Operation *materialize(Operation *src, LinkState &state) const override {
+    return state.clone(src);
   }
 
   SmallVector<Operation *> dependencies(Operation *op) const override {
+    // TODO: use something like SymbolTableAnalysis
+    Operation *module = op->getParentOfType<ModuleOp>();
+    SymbolTable st(module);
     SmallVector<Operation *> result;
-
-    Operation *symbolTableOp = symbolTable->getOp();
     op->walk([&](SymbolUserOpInterface user) {
       if (user.getOperation() == op)
         return;
 
-      if (SymbolRefAttr symbol = user.getUserSymbol())
-        if (Operation *dep = symbolTable->lookupSymbolIn(symbolTableOp, symbol))
+      if (SymbolRefAttr symbol = user.getUserSymbol()) {
+        if (Operation *dep = st.lookup(symbol.getRootReference())) {
           result.push_back(dep);
+        }
+      }
     });
 
     return result;
   }
 
 private:
-  std::unique_ptr<SymbolTable> symbolTable;
+  StringAttr getUniqueNameIn(SymbolTable &st, StringAttr name) const {
+    MLIRContext *context = name.getContext();
+    int uniqueId = 0;
+    Twine prefix = name.getValue() + ".";
+    while (st.lookup(name))
+      name = StringAttr::get(context, prefix + Twine(uniqueId++));
+    return name;
+  }
 
-  SetVector<Operation *> valuesToClone;
+  void renameSymbolRefIn(Operation *op, StringAttr newName) const {
+    AttrTypeReplacer replacer;
+    replacer.addReplacement([&](SymbolRefAttr attr) {
+      return SymbolRefAttr::get(newName, attr.getNestedReferences());
+    });
+    replacer.replaceElementsIn(op);
+  }
+
+  LogicalResult renameRemappedUsersOf(Operation *op, StringAttr newName,
+                                      LinkState &state) const {
+    ModuleOp module = op->getParentOfType<ModuleOp>();
+    // TODO: use something like SymbolTableAnalysis
+    SymbolTable src(module);
+    if (auto uses = src.getSymbolUses(op, module)) {
+      for (SymbolTable::SymbolUse use : *uses) {
+        // TODO: add test where user is not remapped
+        Operation *dstUser = state.remapped(use.getUser());
+        if (!dstUser)
+          continue;
+        renameSymbolRefIn(dstUser, newName);
```

Comment:
minor nit: Isn't the repeated initialization of the replacer (in `renameSymbolRefIn`) kind of costly?

Reply:
It needs to be initialized with `newName` and also it is just emplace_back of a single function pointer, no biggie I think.